### PR TITLE
Ampers - Lily Sky

### PR DESCRIPTION
**Pull request by lillers1122**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | My design decision that took the longest was probably choosing to have a block class as a child of a reservation class. I considered making reservations and blocks separate classes, but because both blocks and reservations have start_dates, end_dates, and costs associated with them it felt reasonable to have the blocks inherit those instance variables... and then have their own block_ids and block_rooms. If I was to redo this project I would do away with a room class and just have frontdesk, reservation, and block as three distinct classes in the hotel module.
Describe a concept that you gained more clarity on as you worked on this assignment. | This assignment helped me gain clarity on scope: specifically using methods by giving "windows" in with () and returning desired data (including returning multiple things at one time - yay for little victories).
Describe a _nominal test_ that you wrote for this assignment. | Loading static CSV data and checking to make sure the expected number of reservations loaded is a nominal test: <br><br> concierge = Hotel::FrontDesk.new('support/test_reservations.csv', 'support/test_blocks.csv') <br> concierge.reservations.length.must_equal 3
Describe an _edge case test_ that you wrote for this assignment. | Trying to reserve a room from a block when all the rooms in that block are already booked requires an edge case test: <br><br> @concierge = Hotel::FrontDesk.new('support/test_reservations.csv', 'support/test_blocks.csv') <br> @concierge.blocks.length.must_equal 1 <br> @concierge.make_room_block(3, 160, '2018-5-5', '2018-5-7') <br> it "raises an error if no rooms are available" do<br> proc {<br>      4.times do<br>       @concierge.book_blocked_room(2)<br>       end }.must_raise ArgumentError<br> end
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I think I did okay with pseudocode. I treated my test writing as my pseudocode more often than not.

### CODE REVIEW

File: concierge.rb [hunk 1]
```diff
@@ -0,0 +1,392 @@
+#Lily Sky | Ada c9 Ampers | Hotel Project, Ruby
```
Comment (CheezItMan) [inline concierge.rb:1]: `concierge`!  Fancy!

File: lib/front_desk.rb [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+      new_block = Hotel::Block.new({
+        block_id: @blocks.length + 1,
+        block_rooms: rooms,
+        start_date: r_start,
+        end_date: r_end,
+        cost: cost
+        })
+
+      @blocks << new_block
+      room_block_reservations(new_block)
```
Comment (CheezItMan) [inline lib/front_desk.rb:176]: Usually you don't make reservations for a block immediately.  Instead the rooms are reserved for reservations later.

File: lib/front_desk.rb [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+          end_date: block.end_date.to_s,
+          block_status: :AVAILABLE,
+          cost: block.cost
+          })
+        @reservations << new_reservation
+      end
+    end
+
+    def find_reservations_in_block(r_block_id)
+      @reservations.select { |reservation| reservation.block_id != nil && reservation.block_id == r_block_id}
```
Comment (CheezItMan) [inline lib/front_desk.rb:196]: So reservations have to know about blocks (block_id)?

File: lib/reservation.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    def duration
+      (@end_date - @start_date).to_i
+    end
+
+    def projected_cost
+      charge = (duration * @cost).to_f.round(2)
+    end
+
+    def overlap?(r_start, r_end)
```
Comment (CheezItMan) [inline lib/reservation.rb:32]: 👍

File: lib/room.rb [hunk 5]
```diff
@@ -0,0 +1,13 @@
+#Lily Sky | Ada c9 Ampers | Hotel Project, Ruby
+require 'date'
+
+module Hotel
```
Comment (CheezItMan) [inline lib/room.rb:4]: Is this class needed?

**Comment by CheezItMan** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Design** |
Demonstrated classes having a single responsibility |  Check
Demonstrated loose coupling |  `Reservation` and `Block` are tightly coupled.  
Methods demonstrate a good use of encapsulation, inputs and outputs | Check
Wave 1 requirements |  Check
Wave 2 requirements |  Check
Wave 3 requirements |  Nice work making `Block` a subclass of `Reservation`.  
Summary | Excellent work on the extras!  You hit the learning objectives, and check out my comments.  
 Nice work on the terminal interface!  You can also check out our [reference implementation](https://github.com/droberts-ada/hotel/tree/dpr/solution).

File: specs/front_desk_spec.rb [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+
+    it "must raise an error if no rooms are available" do
+      proc {
+      @concierge.reserve_room('2018-5-5','2018-5-7')
+      @concierge.rooms_available('2018-5-5','2018-5-7')}.must_raise ArgumentError
+    end
+  end
+
+  describe "reserve_room method" do
+    before do
```
Comment (CheezItMan) [inline specs/front_desk_spec.rb:113]: More testing is needed here.
